projectq/backends/_azure/_azure_quantum.py

```diff
@@ -0,0 +1,403 @@
+from projectq.ops import (
+    AllocateQubitGate,
+    DeallocateQubitGate,
+    FlushGate,
+    MeasureGate
+)
+
+from ._azure_quantum_client import send, retrieve
+from ._util import (
+    IONQ_PROVIDER_ID,
+    QUANTINUUM_PROVIDER_ID,
+    is_available_ionq,
+    is_available_quantinuum,
+    to_json,
+    to_qasm,
+    rearrange_result
+)
+
+from ._exceptions import AzureQuantumTargetNotFoundError
+
+try:
+    from azure.quantum import Workspace
+    from azure.quantum.target import Target, IonQ, Quantinuum
+    from azure.quantum.target.target_factory import TargetFactory
+except ImportError:  # pragma: no cover
+    raise ImportError(
+        "Missing optional 'azure-quantum' dependencies. To install run: pip install projectq[azure-quantum]"
+    )
+
+
+class AzureQuantumBackend(BasicEngine):  # pylint: disable=too-many-instance-attributes
+    """Backend for building circuits and submitting them to the Azure Quantum."""
+
+    DEFAULT_TARGETS = {
+        IONQ_PROVIDER_ID: IonQ,
+        QUANTINUUM_PROVIDER_ID: Quantinuum
+    }
+
+    def __init__(
+        self,
+        use_hardware=False,
+        num_runs=100,
+        verbose=False,
+        workspace=None,
+        target_name='ionq.simulator',
+        num_retries=100,
+        interval=1,
+        retrieve_execution=None,
+        **kwargs
+    ):  # pylint: disable=too-many-arguments
+        """
+        Initialize an Azure Quantum Backend object.
+
+        Args:
+            use_hardware (bool, optional): Whether or not to use real hardware or just a simulator. If False,
+            regardless of the value of ```target_name```, ```ionq.simulator``` used for IonQ provider and
+                ```quantinuum.hqs-lt-s1-apival``` used for Quantinuum provider. Defaults to False.
+            num_runs (int, optional): Number of times to run circuits. Defaults to 100.
+            verbose (bool, optional): If True, print statistics after job results have been collected. Defaults to
+                False.
+            workspace (Workspace, optional): Azure Quantum workspace. If workspace is None, kwargs will be used to
+                create Workspace object.
+            target_name (str, optional): Target to run jobs on. Defaults to ```ionq.simulator```.
+            num_retries (int, optional): Number of times to retry fetching a job after it has been submitted. Defaults
+                to 100.
+            interval (int, optional): Number of seconds to wait in between result fetch retries. Defaults to 1.
+            retrieve_execution (str, optional): An Azure Quantum Job ID. If provided, a job result with this ID will be
+                fetched. Defaults to None.
+        """
+        super().__init__()
+
+        if target_name in IonQ.target_names:
+            self._provider_id = IONQ_PROVIDER_ID
+        elif target_name in Quantinuum.target_names:
+            self._provider_id = QUANTINUUM_PROVIDER_ID
+        else:  # pragma: no cover
```

Comment:
This `#pragma` is really unneeded since a very simple test case with a random string is able to cover it.

Reply:
This is updated, I already have testcase covers invalid `target_name`, so we should be good here.